powerscale/provider/s3_key_resource.go

```diff
@@ -0,0 +1,253 @@
+			Computed:            true,
+			MarkdownDescription: "The secret key of the key. Computed.",
+			Description:         "The secret key of the key. Computed.",
+		},
+		"secret_key_timestamp": schema.Int64Attribute{
+			Computed:            true,
+			MarkdownDescription: "The timestamp of the secret key. Computed.",
+			Description:         "The timestamp of the secret key. Computed.",
+		},
+		"old_secret_key": schema.StringAttribute{
+			Computed:            true,
+			MarkdownDescription: "The secret key of the old key. Computed.",
+			Description:         "The secret key of the old key. Computed.",
+		},
+		"old_key_expiry": schema.Int64Attribute{
+			Computed:            true,
+			MarkdownDescription: "The expiry of the old key. Computed.",
+			Description:         "The expiry of the old key. Computed.",
+		},
+		"old_key_timestamp": schema.Int64Attribute{
+			Computed:            true,
+			MarkdownDescription: "The timestamp of the old key. Computed.",
+			Description:         "The timestamp of the old key. Computed.",
+		},
+	}
+}
+
+// Create allocates the resource.
+func (r *S3KeyResource) Create(ctx context.Context, request resource.CreateRequest, response *resource.CreateResponse) {
+	var s3key models.S3KeyResourceData
+	diags := request.Plan.Get(ctx, &s3key)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+	// call create s3key
+	resp, err := helper.GenerateS3Key(ctx, r.client, s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error creating s3 key ", err.Error())
+		return
+	}
+	err = helper.CopyFieldsToNonNestedModel(ctx, resp.Keys, &s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error creating s3 key ", err.Error())
+		return
+	}
+	response.State.Set(ctx, s3key)
+}
+
+// Read reads data from the resource.
+func (r *S3KeyResource) Read(ctx context.Context, request resource.ReadRequest, response *resource.ReadResponse) {
+
+	var s3key models.S3KeyResourceData
+	diags := request.State.Get(ctx, &s3key)
+	response.Diagnostics.Append(diags...)
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	// call get s3key
+	resp, err := helper.GetS3Key(ctx, r.client, s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error reading s3 key ", err.Error())
+		return
+	}
+
+	// precheck to invalidate the refresh
+	errMsg := "[UNKNOWN KEY] Key Generated Outside of Terraform"
+	if resp.Keys.GetOldKeyTimestamp() != int32(s3key.SecretKeyTimestamp.ValueInt64()) {
+		s3key.OldSecretKey = types.StringValue(errMsg)
+	} else {
+		s3key.OldSecretKey = s3key.SecretKey
+	}
+	if resp.Keys.GetSecretKeyTimestamp() != int32(s3key.SecretKeyTimestamp.ValueInt64()) {
+		response.Diagnostics.AddWarning(errMsg, errMsg)
+		s3key.SecretKey = types.StringValue(errMsg)
+	}
+	err = helper.CopyFieldsToNonNestedModel(ctx, resp.Keys, &s3key)
+	if err != nil {
+		response.Diagnostics.AddError("Error reading s3 key ", err.Error())
+	}
+	response.State.Set(ctx, s3key)
```

Comment:
What if it fails to set the state?

Reply:
added the return inside if error block

Reply:
response.Diagnostics.Append(response.State.Set(ctx, s3key)) ?

Reply:
added